File: x/wasmx/keeper/testdata/python/forward.py
import json
from wasmx import call, call_static, bech32_string_to_bytes, log, storage_store, storage_load, log

def instantiate():
    store("python")

def main(input):
    print("--py-main", input)
    if "forward" in input:
        return forward(*input["forward"])
    if "forward_get" in input:
        return forward_get(*input["forward_get"])
    raise ValueError('Invalid function')

def forward(value, addresses):
    value = value + load().decode()
    print("--py-forward", value, addresses)
    doLog(value.encode(), [])

    if len(addresses) == 0:
        return value.encode()

    value = value + " -> "
    addressbech32 = addresses.pop(0)
    print("--py-forward call: ", addressbech32, addresses)
    calldata = json.dumps({"forward":[value, addresses]})
    print("--py-forward call data: ", calldata)
    address = bech32_string_to_bytes(addressbech32)
    res = call(1000000, address, 0, calldata.encode())
    response = json.loads(res.decode())
    print("--py-success", response["success"])
    if response["success"] != 0:
        raise ValueError('[py] call failed')
    return bytes(response["data"])

def forward_get(addresses):
    if len(addresses) == 0:
        return load()
    addressbech32 = addresses.pop(0)
    calldata = json.dumps({"forward_get":[addresses]})
    address = bech32_string_to_bytes(addressbech32)
    res = call_static(1000000, address, calldata.encode())
    response = json.loads(res)
    if response["success"] != 0:
        raise ValueError('[py] call_static failed')
    data = bytes(response["data"])
    return load() + " -> ".encode() + data

def store(a):
    value = a.encode()
    key = "key".encode()
    storage_store(key, value)

def load():
    key = "key".encode()
    value = storage_load(key)
    return value

def doLog(databz: bytes, topicsbz: list[bytes]):
    data = [x for x in databz]
    topics = [[y for y in x] for x in topicsbz]
    logdata = json.dumps({"data": data,"topics": topics})
    print("--py-forward logdata-", logdata)
    log(logdata.encode())

File: tests/testdata/js/forward.js
import * as wasmx from 'wasmx';

export function instantiate() {
    store("javascript");
}

export function main(dataObj) {
    if (dataObj.forward) {
        return forward(...dataObj.forward);
    } else if (dataObj.forward_get) {
        return forward_get(...dataObj.forward_get);
    }
    throw new Error("invalid function");
}

function forward(value, addresses) {
    value = value + arrayBufferToString(load())
    doLog(stringToArrayBuffer(value), []);

    if (addresses.length == 0) {
        return stringToArrayBuffer(value);
    }
    value = value + " -> "
    let addressbech32 = addresses.shift();
    let calldata = JSON.stringify({"forward":[value, addresses]})
    let address = wasmx.bech32StringToBytes(addressbech32)
    let res = wasmx.call(1000000, address, new ArrayBuffer(32), stringToArrayBuffer(calldata));
    let response = JSON.parse(arrayBufferToString(res));
    if (response.success != 0) {
        throw new Error("[js] call failed");
    }
    let data = new Uint8Array(Object.values(response.data));
    return data.buffer;
}

function forward_get(addresses) {
    if (addresses.length == 0) {
        return load();
    }
    let addressbech32 = addresses.shift();
    let calldata = JSON.stringify({"forward_get":[addresses]})
    let address = wasmx.bech32StringToBytes(addressbech32)
    let res = wasmx.callStatic(1000000, address, stringToArrayBuffer(calldata))
    let response = JSON.parse(arrayBufferToString(res))
    if (response.success != 0) {
        throw new Error("[js] call_static failed");
    }
    let data = new Uint8Array([...new Uint8Array(load()), ...new Uint8Array(stringToArrayBuffer(" -> ")), ...response.data]);
    return data.buffer;
}

function store(value_) {
    const key = stringToArrayBuffer("key");
    const value = stringToArrayBuffer(value_);
    wasmx.storageStore(key, value);
}

function load() {
    const key = stringToArrayBuffer("key");
    return wasmx.storageLoad(key);
}

// utils
function stringToArrayBuffer(inputString) {
    const bytes = new Uint8Array(inputString.length);
    for (let i = 0; i < inputString.length; i++) {
        bytes[i] = inputString.charCodeAt(i) & 0xFF;
    }
    return bytes.buffer;
}

function arrayBufferToString(arrayBuffer) {
    const bytes = new Uint8Array(arrayBuffer);
    let result = "";
    for (let i = 0; i < bytes.length; i++) {
        result += String.fromCharCode(bytes[i]);
    }
    return result;
}

// data: ArrayBuffer, topics: []ArrayBuffer
function doLog(databz, topics) {
    let logdata = JSON.stringify({"data": [...new Uint8Array(databz)],"topics":[]})
    wasmx.log(stringToArrayBuffer(logdata))
}
